File: src/main/java/io/veridia/jsonlogic/PathResolver.java
package io.veridia.jsonlogic;

import java.util.List;
import java.util.Map;

public class PathResolver {

    /**
     * Splits a dotted variable path into its segments once, at compile time.
     * Kept out of the eval hot path so callers never pay for regex parsing per evaluation.
     */
    public static String[] split(String path) {
        if (path == null || path.isEmpty()) return new String[0];
        return path.split("\\.");
    }

    /**
     * Resolves a dotted path against a root object.
     * <p>
     * Prefer the {@link #resolve(Object, String[])} overload on hot paths: it accepts an
     * already-split path so no regex splitting happens per evaluation.
     */
    public static Object resolve(Object root, String path) {
        if (root == null) return null;
        if (path == null || path.isEmpty()) return root;
        return resolve(root, split(path));
    }

    /**
     * Resolves a pre-split path against a root object. This is the allocation-free path used
     * by compiled {@code var} nodes.
     */
    @SuppressWarnings("unchecked")
    public static Object resolve(Object root, String[] parts) {
        if (root == null) return null;
        if (parts == null || parts.length == 0) return root;

        Object current = root;

        for (String part : parts) {
            if (current == null) return null;

            // Map access
            if (current instanceof Map) {
                current = ((Map<String, Object>) current).get(part);
                continue;
            }

            // List access (array index)
            if (current instanceof List) {
                int index = parseIndex(part);
                if (index < 0) return null;
                List<?> list = (List<?>) current;
                if (index >= list.size()) return null;
                current = list.get(index);
                continue;
            }

            if (current.getClass().isArray()) {
                int index = parseIndex(part);
                if (index < 0) return null;
                Object[] list = (Object[]) current;
                if (index >= list.length) return null;
                current = list[index];
                continue;
            }

            // Nothing else supports deep navigation
            return null;
        }

        return current;
    }

    /**
     * Resolves a single, dot-free key against a root object. Fast path for flat variable access,
     * which is by far the most common {@code var} shape.
     */
    @SuppressWarnings("unchecked")
    public static Object resolveSingle(Object root, String key) {
        if (root == null) return null;

        if (root instanceof Map) {
            return ((Map<String, Object>) root).get(key);
        }

        if (root instanceof List) {
            int index = parseIndex(key);
            if (index < 0) return null;
            List<?> list = (List<?>) root;
            return index < list.size() ? list.get(index) : null;
        }

        if (root.getClass().isArray()) {
            int index = parseIndex(key);
            if (index < 0) return null;
            Object[] list = (Object[]) root;
            return index < list.length ? list[index] : null;
        }

        return null;
    }

    /**
     * Parses a path segment as a non-negative array index without throwing.
     * Returns {@code -1} for anything that is not a valid non-negative integer.
     */
    private static int parseIndex(String part) {
        if (part.isEmpty()) return -1;
        int value = 0;
        for (int i = 0; i < part.length(); i++) {
            char c = part.charAt(i);
            if (c < '0' || c > '9') return -1;
            value = value * 10 + (c - '0');
            if (value < 0) return -1; // overflow
        }
        return value;
    }
}

File: src/main/scala/io/veridia/jsonlogic/PathResolver.scala
package io.veridia.jsonlogic

/** Resolves a `var` path against a root context object. Walks only `java.util.Map`/
  * `java.util.List`/`Object[]` — never a POJO's fields, matching the compiler's dynamic,
  * duck-typed evaluation model.
  */
object PathResolver:

  def split(path: String): Array[String] =
    if path == null || path.isEmpty then Array.empty else path.split("\\.")

  def resolve(root: Any, path: String): Any =
    if root == null then null
    else if path == null || path.isEmpty then root
    else resolve(root, split(path))

  def resolve(root: Any, parts: Array[String]): Any =
    if root == null then null
    else if parts == null || parts.isEmpty then root
    else
      var current: Any = root
      var i = 0
      while current != null && i < parts.length do
        current = step(current, parts(i))
        i += 1
      current

  def resolveSingle(root: Any, key: String): Any =
    if root == null then null else step(root, key)

  private def step(current: Any, part: String): Any = current match
    case m: java.util.Map[?, ?] => m.asInstanceOf[java.util.Map[String, AnyRef]].get(part)
    case l: java.util.List[?] =>
      val idx = parseIndex(part)
      if idx < 0 || idx >= l.size then null else l.get(idx)
    case a: Array[AnyRef] =>
      val idx = parseIndex(part)
      if idx < 0 || idx >= a.length then null else a(idx)
    case _ => null

  /** Parses a non-negative array index without throwing; `-1` for anything else. */
  private def parseIndex(part: String): Int =
    if part.isEmpty then -1
    else
      var value = 0
      var i = 0
      var ok = true
      while ok && i < part.length do
        val c = part.charAt(i)
        if c < '0' || c > '9' then ok = false
        else
          val digit = c - '0'
          if value > (Int.MaxValue - digit) / 10 then ok = false else value = value * 10 + digit
        i += 1
      if ok then value else -1
